### PR TITLE
Adding an abstract consensus spec

### PR DESCRIPTION
**Pull request by heidihoward**
Adding a simple abstract consensus spec

### CODE REVIEW

File: tla/consensus/MCccfraft.tla [hunk 1]
```diff
@@ -126,4 +126,16 @@ DebugNotTooManySigsInv ==
     \A i \in Servers:
         FoldSeq(LAMBDA e, count: IF e.contentType = TypeSignature THEN count + 1 ELSE count, 0, log[i]) < 8
 
+----
+
+MappingToAbs == 
+  INSTANCE abs WITH
+    Servers <- Servers,
+    Terms <- 1..MaxTermLimit,
+    MaxLogLength <- 4 + (RequestLimit*2) + MaxTermLimit,
```
Comment (lemmy) [inline tla/consensus/MCccfraft.tla:135]: `+4` because of the max length of the initial logs?

File: tla/consensus/abs.tla [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+Copy(i) ==
+    \E j \in Servers : 
+        /\ Len(CLogs[j]) > Len(CLogs[i])
+        /\ \E l \in 1..(Len(CLogs[j]) - Len(CLogs[i])) : 
+                CLogs' = [CLogs EXCEPT ![i] = CLogs[i] \o SubSeq(CLogs[j], Len(CLogs[i]) + 1, Len(CLogs[i]) + l)]
+
+\* The node with the longest log can extend its log.
+Extend(i) ==
+    /\ \A j \in Servers : Len(CLogs[j]) \leq Len(CLogs[i])
+    /\ \E l \in 0..(MaxLogLength - Len(CLogs[i])) :
```
Comment (lemmy) [inline tla/consensus/abs.tla:32]: Given that it's state-space is small, I'd suggest to conceptually move `MaxLogLength` into a state constraint to improve the clarity of this spec.

File: tla/consensus/abs.tla [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+    
+Init ==
+    CLogs \in [Servers -> InitialLogs]
+
+\* A node i can copy a ledger suffix from another node j.
+Copy(i) ==
+    \E j \in Servers : 
+        /\ Len(CLogs[j]) > Len(CLogs[i])
+        /\ \E l \in 1..(Len(CLogs[j]) - Len(CLogs[i])) : 
+                CLogs' = [CLogs EXCEPT ![i] = CLogs[i] \o SubSeq(CLogs[j], Len(CLogs[i]) + 1, Len(CLogs[i]) + l)]
```
Comment (lemmy) [inline tla/consensus/abs.tla:27]: Substituting `@` for `CLogs[i]` could make it easier to discern `CLogs[j]` from `CLogs[i]`.

File: tla/consensus/abs.tla [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+\* Abstract specification for a distributed consensus algorithm.
+\* Assumes that any node can inspect the state of all other nodes. 
+
+EXTENDS Sequences, SequencesExt, Naturals, FiniteSets
+
+CONSTANT Servers, Terms, MaxLogLength, StartTerm
+
+\* Commit logs from each node
+\* Each log is append-only and the logs will never diverge.
+VARIABLE CLogs
```
Comment (lemmy) [inline tla/consensus/abs.tla:11]: This human linter would have lower-cased the variable name.

File: tla/consensus/abs.tla [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+\* The node with the longest log can extend its log.
+Extend(i) ==
+    /\ \A j \in Servers : Len(CLogs[j]) \leq Len(CLogs[i])
+    /\ \E l \in 0..(MaxLogLength - Len(CLogs[i])) : 
+        \E s \in [1..l -> Terms] :
+            CLogs' = [CLogs EXCEPT ![i] = CLogs[i] \o s]
+
+\* The only possible actions are to append log entries.
+\* By construction there cannot be any conflicting log entries
+\* Log entries are copied if the node is not the longest.
```
Comment (lemmy) [inline tla/consensus/abs.tla:38]: "if the node's log is not the longest"

File: tla/consensus/abs.tla [hunk 6]
```diff
@@ -0,0 +1,58 @@
+---- MODULE abs ----
+\* Abstract specification for a distributed consensus algorithm.
+\* Assumes that any node can inspect the state of all other nodes.
```
Comment (lemmy) [inline tla/consensus/abs.tla:3]: Perhaps, reword to something like "can atomically inspect the state of all other nodes"?

**Review by lemmy — APPROVED** [review]

File: tla/consensus/abs.tla (hunk 4 above)
Comment (achamayou) [reply tla/consensus/abs.tla:11]: That does seem to be the de facto convention in our specs, e.g. https://github.com/microsoft/CCF/blob/main/tla/consensus/ccfraft.tla#L241, https://github.com/microsoft/CCF/blob/main/tla/consistency/Consistency.tla#L135

camelCase for VARIABLES, and CamelCase for CONSTANTS.

File: tla/consensus/abs.tla (hunk 6 above)
Comment (heidihoward) [reply tla/consensus/abs.tla]: https://github.com/microsoft/CCF/pull/6438/commits/1d131f6eeb43042f48cada238fe7474957f843ac

File: tla/consensus/abs.tla (hunk 5 above)
Comment (heidihoward) [reply tla/consensus/abs.tla]: https://github.com/microsoft/CCF/pull/6438/commits/a696ad56bd58ed0ff910323b8cd9700c7ddfb819

File: tla/consensus/abs.tla (hunk 4 above)
Comment (heidihoward) [reply tla/consensus/abs.tla:11]: I have no strong opinion on this but would prefer to keep all specs constant

File: tla/consensus/abs.tla (hunk 3 above)
Comment (heidihoward) [reply tla/consensus/abs.tla]: https://github.com/microsoft/CCF/pull/6438/commits/e9c2fe0af9440f926dee393d34049b5439caf156

File: tla/consensus/MCccfraft.tla (hunk 1 above)
Comment (heidihoward) [reply tla/consensus/MCccfraft.tla]: https://github.com/microsoft/CCF/pull/6438/commits/cf707d278e8e523152ebcaf52b6c39f188d95f7c

**Review by achamayou — APPROVED** [review]

File: tla/consensus/abs.tla (hunk 4 above)
Comment (achamayou) [reply tla/consensus/abs.tla:11]: commitLogs would be consistent with other specs

Comment (heidihoward) [reply tla/consensus/abs.tla]: https://github.com/microsoft/CCF/pull/6438/commits/2ab3dfec0eb1a1be140e845624a60f9c5d992707

File: tla/consensus/abs.tla (hunk 2 above)
Comment (heidihoward) [reply tla/consensus/abs.tla]: Is this (https://github.com/microsoft/CCF/pull/6438/commits/8931e18313570a194edaf7a648d434f8556f283b) what you mean or did you mean something different?

**Pull request merged**